### ISSUE TITLE
[catalog] Users should get clear information about refresh loop warnings/errors

### ISSUE THREAD

**Issue by freben** [spotify/backstage#2876]
## Feature Suggestion

As the entity refresh loop runs, a lot of information is emitted as logs - for example, if entities fail validation or if the source location could not be read. This information is not made trivially available for the end user, which makes it very hard to fix a problem e.g. with a syntax error in your YAML file.

All refresh loop information that is relevant to end users should be viewable directly in the catalog interface. At the very least, on an entity overview page you should see any warnings and errors related to your entity declaration. It would also be helpful to be able to see a clear status marker of the high level state when looking at entities in a list view (for example, an orange warning marker telling you that something needs your attention).

## Possible Implementation

For the catalog frontend plugin, it should show warning/error info on the entity overview pages.

For the catalog backend plugin, it should gather up information as it processes entities and persist that information to the database as well as writing to logs. The information should be tagged with the current location and the current entity, where available, so that it can easily be queried for after the fact.

The catalog `/entities` and `/locations` APIs should emit at least some summary of this status information as part of their output.

## Future extension

Additional APIs could be added to extract more detailed information, such as all log entries relevant to the refresh run, in a future addition.

Additional APIs could be added to let users "dry run" an ingestion which goes through the full processing loop but does not write out the entities. This can then be used for PR feedback and other validation tools.

## Context

Entity declaration files are a major contact surface for users, and the "asynchronous black box" of refreshes must expose enough information for users to be effective.

Related issues:

- #2292 describes a status API whose backend part could be the foundation for the storage of this information
- #2053 discusses these same issues

**Comment by stefanalund** [comment]
One option is to use this UX:

![Screenshot 2020-10-15 at 15 52 01](https://user-images.githubusercontent.com/190856/96138784-6aba2c00-0efe-11eb-8141-f9553bbab7e3.png)

If the status is not green, maybe we show a "More details" button that pops up a dialog with more details. @katz95 any thoughts?

**Comment by dtuite** [comment]
@freben are you actively working on this? We have a new team member joining today who could pick this up.

**Comment by Fox32** [comment]
I played a little bit around with 

> Additional APIs could be added to let users "dry run" an ... This can then be used for PR feedback and other validation tools.

PR feedback is something we are looking for. I think this would be quite easy to implement. I did my experiments here: https://github.com/spotify/backstage/compare/master...SDA-SE:dry-run-experiment
If that goes into the right direction, I can clean it up and make it a real PR.

> The catalog /entities and /locations APIs should emit at least some summary of this status information as part of their output.

That is an issue right now. If it's not successful a play exception is returned (just a string). But that would also be easy to improve in a follow-up PR.

**Comment by freben** [comment]
@Fox32 that looks sensible!

**Comment by Fox32** [comment]
@freben I made a PR for the dry run part: #3013

**Comment by Fox32** [comment]
I hijacked the issue a bit with the dry run 😉 Let's get back to original topic. Showing the current status of ingestion. I looked a bit around what already is in place:

* Right now `/api/catalog/locations` already contains `{"currentStatus":{"message":null,"status":"success","timestamp":"2020-11-25 10:12:27"}` in its response. For now it isn't working correctly. I couldn't make it display failures, for example for syntax errors.
* Syntax errors are logged [here](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts#L153-L157) but don't seem to cause a call to `locationsCatalog.logUpdateFailure` therefore they aren't included. If I fix that, locations are still always set to success afterwards [here](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts#L130).
* I would recommend to fix that be moving all status related calls into `refreshSingleLocation`. I fix that, I still have the issue that the status is not only logged for every location, but also every entity [here](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts#L185-L190). The successful entities override the broken location. I can remove that too and have a result: `"currentStatus":{"message":"YAML error, YAMLSyntaxError: All collection items must start at the same column","status":"fail","timestamp":"2020-11-25 10:36:31"}`.
* The `currentStatus` only returns the [last logged status](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/CommonDatabase.ts#L384). I wonder if we should include some kind of "run-id" (or timestamp of the run) that is used to collect all status details from a run of a location and return them together in `/api/catalog/locations`.
* If we extend `/api/catalog/entities` with the `currentStatus`, we can filter the log by entity and include the entity specific status. Right now it only contains the entity name, should it be the `uid` instead? Entity names are only unique per kind? We might also include the logs for the location here. Syntax errors only affect the location and not the entity. They will cause the entity ingestion to be skipped without an error.

I would like to have some opinions on that and could create a PR.

To show the status in the about card like Stefan suggested, we have two options:
1. Parse `managed-by-location` in frontend and retrieve the status of a location. Additional call, but is possible without changing the backend.
2. Extend the `/api/catalog/entities` endpoint to resolve the `currentStatus` in the backend. Probably the long term solution,  as parsing `managed-by-location` feels strange.

Beside that I would go with the suggested UI from Stefan. If it's not green, the use can select "Show More" and sees the error message in a dialog.

**Comment by freben** [comment]
@Fox32 Great input!

I've been thinking about adding a dedicated status table in the catalog. Gonna brain dump a bit before bed and see if you think it aligns with what you observed.

Rough sketch:

```sql
CREATE TABLE status (
  t        TIMESTAMP NOT NULL,
  source   TEXT NOT NULL,    -- owner of the status, e.g. "catalog-refresh-loop"
  status   TEXT NOT NULL,    -- "pending", "ok", "failed"?
  job_id   TEXT NOT NULL,    -- random for each update, used for associating with events
  -- associative columns
  -- exactly one of these is set
  entity   TEXT NULL,        -- kind:namespace/name
  location TEXT NULL,        -- url:https://...
);
```

There could be an `events` table as well.

```sql
CREATE TABLE events (
  -- core columns
  t TIMESTAMP NOT NULL,
  source TEXT NOT NULL,     -- originator of the event, e.g. "catalog-refresh-loop"
  type TEXT NOT NULL,       -- what happened, e.g. "start"
  severity TEXT NOT NULL,   -- debug, info, warning, error
  message TEXT NOT NULL,
  traceback TEXT NULL,
  -- associative columns
  -- one or several ones of these are set
  job_id TEXT NULL,         -- the current refresh run id for example
  entity_name TEXT NULL,    -- kind:namespace/name
  entity_uid TEXT NULL,     -- the uid of the entity
  location TEXT NULL,       -- url:https://...
  locationId TEXT NULL,     -- uuid of the location 
  -- more...
);
```

As processes run within the various backends, they would emit events (probably using https://github.com/backstage/backstage/issues/3221) that have a number of optional context keys that get put into the associative columns when present. So a single refresh loop run could first emit a "start" type message only associated with its generated job_id, then a "location_start" type message associated with its job_id and the location that it starts on, then for each entity read, it additionally adds the entity columns as context.

This way, you can get the events for a number of different parts, by selecting on a certain slice of associative columns. You can see the set of refreshes as they happen for a single entity, or what happened for an entire location's all entities.

From an API standpoint, an entity would get a new `status` root field when reading.

```yaml
status:
  catalog-refresh-loop:
    # ...
metadata:
  name: ...
```

There would be a new set of API endpoints for reading events. This is used for debugging and tracking the actual inner ongoings. For a scaffolder run, it will contain the events and the log output of the various parts of the template as they progress.

**Comment by Fox32** [comment]
@freben  The notification for your response is already sitting in my inbox for months... 😆 That is also some nice input.

I don't have anything new progress to add but I noticed that the proposed dialog might also be a good location to host some kind of "entity source" view. But not the actual source that was at the location, but how the entity looks once it was ingested. That would sometimes be nice for debugging. But for now dev tools are fine.

**Comment by freben** [comment]
Yes! I would love to see both the raw ingested data and the finished output.

**Comment by freben** [comment]
Related: https://github.com/backstage/backstage/pull/5654

**Comment by alzafacon** [comment]
My team is interested in getting this feature (showing entity refresh errors) closer to done. Is there a small part of this I could work on? I'm struggling to figure out what has already been done and what is left to work on.

**Comment by freben** [comment]
Hey @alzafacon - this is actually being actively worked on right now by the maintainers!

**Comment by desktophero** [comment]
Hey @freben - 

👋 

Is there a recommended way to track the active work on this particular item? I see `help wanted` label is still on ... how can we help? 😄

**Comment by freben** [comment]
Hm I see we don't have an active ticket for it - we are in a bit of a "tying up the loose ends" state with the catalog to mark it as beta and then GA. I think what's missing now is to make sure that the frontend has a nice, useful way to see errors that happen "up the chain" - for example, if you register a location that is so broken that we can't even parse the YAML, there's even no output entity to attach the error to. So we want to expose enough API functionality, and the associated view updates, to be able to clearly see these issues anyway.

This will need some design and deliberation. I think it's fair to say that at this point it may be hard to step in and contribute to those aspects since the fundamentals need to shift a little to accommodate. I'll actually remove the help wanted label briefly and we'll discuss among ourselves how to maybe make some more targeted tickets out of it.

**Comment by timbonicus** [comment]
#6586 is not a complete solution, but at least exposes some catalog refresh errors on entity pages.

**Comment by Rugvip** [comment]
Related to #6973
